### ISSUE TITLE
[Suggestion] Self closing tags

### ISSUE THREAD

**Issue by stsrki** [aspnet/Blazor#1420]
This is not really an issue, just that it would be good to have. Right now the intellisense cannot detect components that should not have closing tags(or does not need them) e.g. `<div></div>`, but instead they would have a "_self-close_" tag e.g. `<input />` or `<img />`.

Possible use case would be for components that does not have `ChildContent`. For example I have an Icon component and I define the icon name through the `Name `parameter so `ChildContent `is not needed. But every time I use it, the VS intellisense will generate closing tag.

**Current behavior:**
`<Icon Name="user"></Icon>`

**Wanted:**
`<Icon Name="user" />`

Maybe this is not connected to the Blazor team but to the VS team instead, please ignore in that case.

**Comment by SteveSandersonMS** [comment]
I agree that would be nice. Not super urgent, but a good suggestion to have on the backlog!

I don't know exactly which team owns the code that autocompletes closing tags. Pretty sure it's not in Blazor itself, but for whether it's in http://github.com/aspnet/razor or VS itself, either @NTaylorMullen or @rynowak would know best.

**Comment by NTaylorMullen** [comment]
It's possible today. In the generated `TagHelperDescriptor` for the Blazor component if you set the corresponding [`TagMatchingRuleDescriptor.TagStructure`](https://github.com/aspnet/Razor/blob/40080f3637b25ceed9a1d0bff9dbffd4c10513ac/src/Microsoft.AspNetCore.Razor.Language/TagMatchingRuleDescriptor.cs#L20) to [`WithoutEndTag`](https://github.com/aspnet/Razor/blob/master/src/Microsoft.AspNetCore.Razor.Language/TagStructure.cs#L26) then the tooling will pick up on that and provide appropriate auto-complete. 

Just to be clear, this would need to be done by the Blazor framework so there's nothing a user can do today to fix it in their app.

**Comment by SteveSandersonMS** [comment]
That's great to hear, @NTaylorMullen! So hopefully this will be relatively straightforward when we or someone in the community gets round to implementing it.

**Comment by NTaylorMullen** [comment]
> @NTaylorMullen - does this affect the binding? or is it just a hint?

It always binds; however, in the case that you say "WithoutEndTag" and you provide an end-tag, an error gets created.

**Comment by rynowak** [comment]
womp womp

**Comment by NTaylorMullen** [comment]
> womp womp

Ah, so you want it not to create an error and have a default experience of no end-tag I assume?

**Comment by rynowak** [comment]
> default experience of no end-tag I assume?

Blazor creates something similar to the DOM, it's not a meaningful distinction since we're not rendering text directly.

**Comment by isc30** [comment]
Is this ticket still up for grab? If yes, I would gladly open a PR

**Comment by SteveSandersonMS** [comment]
@isc30 Perhaps @NTaylorMullen or @rynowak could clarify, but my interpretation of their conversation above is that we don't currently have the capability in Razor tooling itself to understand that self-closing tag helpers can *also* allow optional child content (i.e., it will show a false error message if you do that). If I'm right about that, then the first PR needs to go into Razor to add some new feature for this.

**Comment by NTaylorMullen** [comment]
> If I'm right about that, then the first PR needs to go into Razor to add some new feature for this.

Correct

**Comment by kant2002** [comment]
@NTaylorMullen does there any work/issue in Razor repository which eventually make that happens ?

**Comment by NTaylorMullen** [comment]
> @NTaylorMullen does there any work/issue in Razor repository which eventually make that happens ?

No there is not
